### PR TITLE
Feature/bigint support

### PR DESCRIPTION
**Pull request by zl730**
# PR Details

## Description

- Add big integer data type.
- basic opc (add, minus, multiply, divide, min, max, sqrt)
- refactor some code

## How Has This Been Tested

unit test passed
## Types of changes

-   [x]   New feature (non-breaking change which adds functionality)

## Checklist

-   [x]   My code follows the code style of this project.
-   [x]   My change requires a change to the documentation.
-   [x]   All new and existing tests passed.

### CODE REVIEW

File: src/main/scala/vsys/blockchain/contract/DataEntry.scala [hunk 1]
```diff
@@ -1,74 +1,47 @@
...
-      case DataType.Boolean => Json.toJson(if (d(0) == 1.toByte) "True" else "False")
-      case DataType.ShortBytes => Json.toJson(Base58.encode(d))
-      case DataType.OpcBlock => Json.toJson("Internal code block")
-      case _ => Json.toJson("Unsupported data type")
-    }
-  }
-
   override def equals(obj: Any): Boolean = obj match {
-    case a: DataEntry => bytes sameElements a.bytes
+    case a: DataEntry => (bytes sameElements a.bytes) && DataType.check(dataType.id.toByte, a.dataType.id.toByte)
```
Comment (ncying) [inline src/main/scala/vsys/blockchain/contract/DataEntry.scala:22]: `bytes sameElements a.bytes` is already included `DataType.check(dataType.id.toByte, a.dataType.id.toByte)`?

File: src/main/scala/vsys/blockchain/contract/DataEntry.scala [hunk 2]
```diff
@@ -1,74 +1,47 @@
...
     case _ => false
   }
 
-  override def hashCode(): Int = java.util.Arrays.hashCode(bytes)
+  override def hashCode(): Int = java.util.Arrays.hashCode(Array(dataType.id.toByte) ++ bytes)
```
Comment (ncying) [inline src/main/scala/vsys/blockchain/contract/DataEntry.scala:26]: same, `lazy val bytes: Array[Byte] = Array(dataType.id.asInstanceOf[Byte]) ++ data`, and data type is already included, why we need to add `dataType` again?

File: src/main/scala/vsys/blockchain/contract/DataEntry.scala (hunk 1 above)
Comment (zl730) [reply src/main/scala/vsys/blockchain/contract/DataEntry.scala:22]: `DataType` byte is not included in bytes of `DataEntry`.

Comment (ncying) [reply src/main/scala/vsys/blockchain/contract/DataEntry.scala:22]: check line 14, it should be, right?

File: src/main/scala/vsys/blockchain/contract/DataEntry.scala (hunk 2 above)
Comment (zl730) [reply src/main/scala/vsys/blockchain/contract/DataEntry.scala:26]: No evidence found here to show that data type is  included.

File: src/main/scala/vsys/blockchain/contract/DataEntry.scala (hunk 1 above)
Comment (zl730) [reply src/main/scala/vsys/blockchain/contract/DataEntry.scala:22]: oh, sorry. i got confused between data and bytes

File: src/main/scala/vsys/blockchain/contract/DataEntry.scala [hunk 3]
```diff
@@ -83,16 +56,17 @@ object DataEntry {
     bytes.headOption match {
       case Some(b) =>
         DataType.fromByte(b) match {
-          case Some(dt: DataType.DataTypeVal) if (dt.lenFixed && checkDataType(bytes.drop(1).take(dt.maxLen), dt)) =>
+          case Some(dt: DataType.DataTypeVal[_]) if (dt.lenFixed && dt.validator(bytes.drop(1).take(dt.maxLen))) =>
             Right((DataEntry(bytes.drop(1).take(dt.maxLen), dt), bytes.drop(1 + dt.maxLen)))
-          case Some(dt: DataType.DataTypeVal) if (!dt.lenFixed && dt.maxLen <= Short.MaxValue && checkDataType(bytes.drop(1).take(2 + Shorts.fromByteArray(bytes.drop(1).take(2))), dt)) =>
-            Right((DataEntry(bytes.drop(1).take(2 + Shorts.fromByteArray(bytes.drop(1).take(2))), dt), bytes.drop(3 + Shorts.fromByteArray(bytes.drop(1).take(2)))))
+          case Some(dt: DataType.DataTypeVal[_]) if (!dt.lenFixed && dt.maxLen <= Short.MaxValue
+            && dt.validator(bytes.drop(1).take(2 + Shorts.fromByteArray(bytes.drop(1).take(2))))) =>
```
Comment (ncying) [inline src/main/scala/vsys/blockchain/contract/DataEntry.scala:62]: I am worried about the `Shorts.fromByteArray` here, if `take(2)` return an empty `Array[Byte]` what will happen here? or we need an extra condition `bytes.length >=3`.

File: src/main/scala/vsys/blockchain/contract/DataType.scala [hunk 4]
```diff
@@ -1,34 +1,162 @@
...
+                                   jsonifier: T => JsValue,
+                                   private val extValidator: Array[Byte] => Boolean) extends Val(dataType) {
+
+    def *(n: Int): Int = n * dataType
+
+    val jsonifierB: Array[Byte] => JsValue = b => jsonifier(deserializer(b))
+
+    // check length and other limits
+    val validator: Array[Byte] => Boolean = data => (((lenFixed && data.length == maxLen)
+        || (!lenFixed && data.length > 2 && Shorts.fromByteArray(data.take(2)) == data.drop(2).length && data.drop(2).length <= maxLen))
```
Comment (ncying) [inline src/main/scala/vsys/blockchain/contract/DataType.scala:32]: How about `data.length == 2` case? e.g. empty shortBytes? may add a related unit test for this boundary case.

File: src/main/scala/vsys/blockchain/contract/DataType.scala [hunk 5]
```diff
@@ -1,34 +1,162 @@
...
+    jsonifier         = a => Json.toJson(a.address),
+    extValidator      = vsys.account.ContractAccount.fromBytes(_).isRight)
+
+  val Account         = DataTypeVal[vsys.account.Account](7,
+    lenFixed          = false,
+    maxLen            = (vsys.account.Address.AddressLength).max(vsys.account.ContractAccount.AddressLength),
+    deserializer      = vsys.account.Account.fromBytes(_, 0).right.get._1,
+    serializer        = a => a.bytes.arr,
+    jsonifier         = a => Json.toJson(a.bytes.arr),
+    extValidator      = vsys.account.Account.fromBytes(_, 0).isRight)
```
Comment (ncying) [inline src/main/scala/vsys/blockchain/contract/DataType.scala:90]: One thing a little bit confused here. `Account` type is not `allowed` as DataEntry passed in. It is a combination of `Address` and `ContractAccount` and which only used in function's input data check. But we can not pass an `Account` type `DataEntry` in previous logic. One simple solution we can use here is set `extValidator = _ => false`

File: src/main/scala/vsys/blockchain/contract/DataType.scala (hunk 4 above)
Comment (zl730) [reply src/main/scala/vsys/blockchain/contract/DataType.scala:32]: Is Array[Byte](0, 0) a valid data?

File: src/main/scala/vsys/blockchain/contract/DataType.scala [hunk 6]
```diff
@@ -1,34 +1,162 @@
...
+
+  val ShortBytes      = DataTypeVal[Array[Byte]](11,
+    lenFixed          = false,
+    maxLen            = MaxShortBytesLength,
+    deserializer      = b => b.drop(2),
+    serializer        = b => arrayShortLengthToByteArray(b) ++ b,
+    jsonifier         = b => Json.toJson(Base58.encode(b)),
+    extValidator      = _ => true)
+
+  val Balance         = DataTypeVal[Long](12,
```
Comment (ncying) [inline src/main/scala/vsys/blockchain/contract/DataType.scala:133]: I think we do not use `Balance` at the current stage, may remove it first. Or put it at last with `extValidator = _ => false`.

File: src/main/scala/vsys/blockchain/contract/DataType.scala (hunk 4 above)
Comment (ncying) [reply src/main/scala/vsys/blockchain/contract/DataType.scala:32]: how to represent a `shortBytes` or `shortText` with zero length? I think `Array[Byte](0, 0)` is a valid data for `shortBytes` and `shortText` but not for `BigInteger`(if BigInt(0) is not an empty array)

but we can set `shortBytes` and `shortText` a non-empty array. will it be too strict?

File: src/main/scala/vsys/blockchain/contract/DataType.scala (hunk 5 above)
Comment (zl730) [reply src/main/scala/vsys/blockchain/contract/DataType.scala:90]: OK

File: src/main/scala/vsys/blockchain/contract/DataType.scala (hunk 6 above)
Comment (zl730) [reply src/main/scala/vsys/blockchain/contract/DataType.scala:133]: OK

File: src/main/scala/vsys/blockchain/state/opcdiffs/BasicOpcDiff.scala [hunk 7]
```diff
@@ -220,14 +89,15 @@ object BasicOpcDiff extends OpcDiffer {
...
-    val Maximum     = basicTypeVal(6, 4, (b, d) => maximum(d(b(1)), d(b(2)), d, b(3)))
+    val Add         = basicTypeVal(1, 4, (b, d) => operation(d(b(1)), d(b(2)), d, b(3), add))
+    val Minus       = basicTypeVal(2, 4, (b, d) => operation(d(b(1)), d(b(2)), d, b(3), minus))
+    val Multiply    = basicTypeVal(3, 4, (b, d) => operation(d(b(1)), d(b(2)), d, b(3), multiply))
+    val Divide      = basicTypeVal(4, 4, (b, d) => operation(d(b(1)), d(b(2)), d, b(3), divide))
+    val Minimum     = basicTypeVal(5, 4, (b, d) => operation(d(b(1)), d(b(2)), d, b(3), minimum))
+    val Maximum     = basicTypeVal(6, 4, (b, d) => operation(d(b(1)), d(b(2)), d, b(3), maximum))
     val Concat      = basicTypeVal(7, 4, (b, d) => concat(d(b(1)), d(b(2)), d, b(3)))
     val ConstantGet = basicTypeVal(8, 2, (b, d) => constantGet(b.slice(1, b.length-1), d, b(b.length-1)))
+    val SqrtBigInt  = basicTypeVal(9, 3, (b, d) => sqrt(d(b(1)), d, b(2)))
```
Comment (ncying) [inline src/main/scala/vsys/blockchain/state/opcdiffs/BasicOpcDiff.scala:100]: Nice simplification. 
1. We need extra OPC which transform the dataEntry between `Amount`, `Int32`, `Timestamp` and `BigInteger `
2. and, or, xor her only for Boolean type
3. comparison OPC needs to support BigInteger in `CompareOpcDiff`. with extra `>`, `<`, `<=`

But we may merge this PR first and add these new features/functions in next PR.

File: src/main/scala/vsys/blockchain/contract/DataType.scala (hunk 4 above)
Comment (zl730) [reply src/main/scala/vsys/blockchain/contract/DataType.scala]: ok, changed to `data.length >= 2`

File: src/main/scala/vsys/blockchain/state/opcdiffs/BasicOpcDiff.scala (hunk 7 above)
Comment (zl730) [reply src/main/scala/vsys/blockchain/state/opcdiffs/BasicOpcDiff.scala:100]: ok, I'll do it in other pr

**Review by ncying — APPROVED** [review]
